core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -17,25 +17,44 @@
 
 package kafka.server
 
-import kafka.message.MessageSet
-import kafka.security.auth.Topic
-import org.apache.kafka.common.metrics.Metrics
-import org.apache.kafka.common.protocol.SecurityProtocol
-import org.apache.kafka.common.TopicPartition
-import kafka.api._
 import kafka.admin.AdminUtils
+import kafka.api._
 import kafka.common._
 import kafka.controller.KafkaController
 import kafka.coordinator.ConsumerCoordinator
 import kafka.log._
+import kafka.message.MessageSet
+import kafka.network.RequestChannel.{Response, Session}
 import kafka.network._
-import kafka.network.RequestChannel.{Session, Response}
-import org.apache.kafka.common.requests.{JoinGroupRequest, JoinGroupResponse, HeartbeatRequest, HeartbeatResponse, ResponseHeader, ResponseSend}
-import kafka.utils.{ZkUtils, ZKGroupTopicDirs, SystemTime, Logging}
-import scala.collection._
+import kafka.security.auth.{Authorizer, ClusterAction, ConsumerGroup, Create, Describe, Operation, Read, Resource, Topic, Write}
+import kafka.utils.{Logging, SystemTime, ZKGroupTopicDirs, ZkUtils}
 import org.I0Itec.zkclient.ZkClient
-import kafka.security.auth.{Authorizer, Read, Write, Create, ClusterAction, Describe, Resource, Topic, Operation, ConsumerGroup}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.errors.UnsupportedVersionException
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.protocol.{ApiKeys, ProtoUtils, SecurityProtocol}
+import org.apache.kafka.common.requests.{HeartbeatRequest, HeartbeatResponse, JoinGroupRequest, JoinGroupResponse, ResponseHeader, ResponseSend}
 
+import scala.collection._
+
+object KafkaApis {
+  //TODO: this method should only use request.header after all the requests are migrated to use client java request class.
+  // For the requests using old scala class, we need to pass in the API version explicitly because the Request.header will
+  // be null. For requests using new java class. the API version will be None.
+  def validateRequestVersion(request: RequestChannel.Request, apiVersion: Option[Short]) {
+    val highestSupportedVersion = ProtoUtils.latestVersion(request.requestId)
+    val requestApiVersion = Option(request.header) match {
+      //requests using new java classes.
+      case Some(header) => header.apiVersion()
+      //requests using old scala classes.
+      case None => apiVersion.get
```

Comment:
In general, `Option.get` is to be avoided because the error message is not particularly helpful. What I meant is something like (feel free to choose a more appropriate message):

`apiVersion.getOrElse(throw new IllegalArgumentException("apiVersion should be defined if request.header is null`"))`